lldb/tools/lldb-dap/lldb-dap.cpp

```diff
@@ -4028,6 +4049,154 @@ void request_disassemble(const llvm::json::Object &request) {
+//   }]
+// },
+// "ReadMemoryArguments": {
+//   "type": "object",
+//   "description": "Arguments for `readMemory` request.",
+//   "properties": {
+//     "memoryReference": {
+//       "type": "string",
+//       "description": "Memory reference to the base location from which data
+//                       should be read."
+//     },
+//     "offset": {
+//       "type": "integer",
+//       "description": "Offset (in bytes) to be applied to the reference
+//                       location before reading data. Can be negative."
+//     },
+//     "count": {
+//       "type": "integer",
+//       "description": "Number of bytes to read at the specified location and
+//                       offset."
+//     }
+//   },
+//   "required": [ "memoryReference", "count" ]
+// },
+// "ReadMemoryResponse": {
+//   "allOf": [ { "$ref": "#/definitions/Response" }, {
+//     "type": "object",
+//     "description": "Response to `readMemory` request.",
+//     "properties": {
+//       "body": {
+//         "type": "object",
+//         "properties": {
+//           "address": {
+//             "type": "string",
+//             "description": "The address of the first byte of data returned.
+//                             Treated as a hex value if prefixed with `0x`, or
+//                             as a decimal value otherwise."
+//           },
+//           "unreadableBytes": {
+//             "type": "integer",
+//             "description": "The number of unreadable bytes encountered after
+//                             the last successfully read byte.\nThis can be
+//                             used to determine the number of bytes that should
+//                             be skipped before a subsequent
+//             `readMemory` request succeeds."
+//           },
+//           "data": {
+//             "type": "string",
+//             "description": "The bytes read from memory, encoded using base64.
+//                             If the decoded length of `data` is less than the
+//                             requested `count` in the original `readMemory`
+//                             request, and `unreadableBytes` is zero or
+//                             omitted, then the client should assume it's
+//                             reached the end of readable memory."
+//           }
+//         },
+//         "required": [ "address" ]
+//       }
+//     }
+//   }]
+// },
+void request_readMemory(const llvm::json::Object &request) {
+  llvm::json::Object response;
+  FillResponse(request, response);
+  auto arguments = request.getObject("arguments");
+
+  lldb::SBProcess process = g_dap.target.GetProcess();
+  if (!process.IsValid()) {
+    response["success"] = false;
+    response["message"] = "No process running";
+    g_dap.SendJSON(llvm::json::Value(std::move(response)));
+    return;
+  }
+
+  auto memoryReference = GetString(arguments, "memoryReference");
+  lldb::addr_t addr;
+  if (memoryReference.consumeInteger(0, addr)) {
```

Comment:
Use `GetUnsigned` instead of manually getting the integer:
```
const addr_t memoryReference = GetUnsigned(arguments, "memoryReference", LLDB_INVALID_ADDRESS);
if (memoryReference == LLDB_INVALID_ADDRESS) {
  response["success"] = false;
  response["message"] =
      "Malformed memory reference: " + memoryReference.str();
  g_dap.SendJSON(llvm::json::Value(std::move(response)));
  return;
}
```

Reply:
To my understanding, `GetUnsigned` works for JSON request similar to `{"memoryReference": 1234 }`. However, memory references are string-typed and not integer-typed. The request we are receiving has the form `{"memoryReference": "0x1234"}`

Reply:
FWIW, I copied used `request_disassemble` as inspiration here

Reply:
Gotcha, that is interesting. I guess they let you encode and decode a `memoryReference` how ever you want then right? I mean, you could return "hello" for a memory reference and they will hand that back to you so you can decode it how you want later? It does make sense and some DSPs have multiple address spaces so  they might want to specify the address space like `"[0]0x1000"` or `"[1]0x1000"` as the first can mean `instruction address space at address 0x1000` and the second can mean `data address space at address 0x1000`. 

If this is the case, then we probably want two functions to encode and decode `memoryReferences` to/fromt strings in JSONUtils:
```
std::string EncodeMemoryReference(lldb::addr_t addr);
lldb::addr_t DecodeMemoryReference(std::string memoryReference);
```
And then use that function here to decode the memory reference.

Reply:
> Gotcha, that is interesting. I guess they let you encode and decode a memoryReference how ever you want then right?

Correct. The memory reference is actually also displayed to the user, highlighted in red in the following screenshot.

![memref](https://github.com/user-attachments/assets/3a5697a3-b668-40ea-879f-b1a27080ddaf)

Also for single-address-space architectures, I could imagine that it would be useful to have memory references like `.rodata + 0x1234` based on the section names or `my_function + 0x1234` based on symbol names. Those memory references would convey more information to the user, and would also be stable across restarts where ASLR might shuffle the sections around. But that's an idea for another day / another commit

Reply:
> we probably want two functions to encode and decode memoryReferences to/fromt strings in JSONUtils

done